flox/core.py

```diff
@@ -550,6 +562,160 @@ def chunk_argreduce(
 
 
 def chunk_reduce(
+    array: np.ndarray,
+    by: np.ndarray,
+    func: T_Funcs,
+    expected_groups: pd.Index | None,
+    axis: T_AxissOpt = None,
+    fill_value: T_FillValues = None,
+    dtype: T_Dtypes = None,
+    reindex: bool = False,
+    engine: T_Engine = "numpy",
+    kwargs: Sequence[dict] | None = None,
+    sort: bool = True,
+) -> IntermediateDict:
+    """
+    Wrapper for numpy_groupies aggregate that supports nD ``array`` and
+    mD ``by``.
+
+    Core groupby reduction using numpy_groupies. Uses ``pandas.factorize`` to factorize
+    ``by``. Offsets the groups if not reducing along all dimensions of ``by``.
+    Always ravels ``by`` to 1D, flattens appropriate dimensions of array.
+
+    When dask arrays are passed to groupby_reduce, this function is called on every
+    block.
+
+    Parameters
+    ----------
+    array : numpy.ndarray
+        Array of values to reduced
+    by : numpy.ndarray
+        Array to group by.
+    func : str or Callable or Sequence[str] or Sequence[Callable]
+        Name of reduction or function, passed to numpy_groupies.
+        Supports multiple reductions.
+    axis : (optional) int or Sequence[int]
+        If None, reduce along all dimensions of array.
+        Else reduce along specified axes.
+
+    Returns
+    -------
+    dict
+    """
+
+    if not (isinstance(func, str) or callable(func)):
+        funcs = func
+    else:
+        funcs = (func,)
+    nfuncs = len(funcs)
+
+    if isinstance(dtype, Sequence):
+        dtypes = dtype
+    else:
+        dtypes = (dtype,) * nfuncs
+    assert len(dtypes) >= nfuncs
```

Comment:
```suggestion
    assert len(dtypes) == nfuncs
```

IMO we should be more strict here. This looseness has resulted in bugs/confusions before

Reply:
Agreed.

Reply:
Ok, I tried it out and some parts of the code actually relies on this behaviour. `>=` was in the original code so I'm going to leave it the same in this PR.